codebase/superdarn/src.bin/tk/tool/fit_median_filter.1.0/fit_median_filter.c

```diff
@@ -0,0 +1,322 @@
+Filtering is performed separately for each beam and channel. 
+Output is a fitacf file.
+
+(C) Copyright 2021 E.C.Bland
+author: E.C.Bland
+
+This file is part of the Radar Software Toolkit (RST).
+
+RST is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <https://www.gnu.org/licenses/>.
+       
+Modifications:
+
+   
+*/
+
+
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/stat.h>
+#include <string.h>
+#include <time.h>
+#include <zlib.h>
+
+#include "rtypes.h"
+#include "rtime.h"
+#include "option.h"
+#include "dmap.h"
+#include "rprm.h"
+#include "fitdata.h"
+
+#include "fitread.h"
+#include "fitwrite.h"
+
+#include "errstr.h"
+#include "hlpstr.h"
+
+#define tmax 2000
+#define maxbm 30
+#define maxch 3
+#define maxrng 250
+
+int fnum=0;
+
+struct RadarParm *prm;
+struct FitData *fit;
+
+struct OptionData opt;
+
+
+int rst_opterr (char *txt) {
+  fprintf(stderr,"Option not recognized: %s\n",txt);
+  fprintf(stderr,"Please try: trim_fit --help\n");
+  return(-1);
+}
+
+
+// function to return the index in an array
+int get_index(int a, int b, int c, int d, int aSize, int bSize, int cSize) {
+    return (d * aSize * bSize * cSize) + (c * aSize * bSize) + (b * aSize) + a; 
+}
+
+// structure to store the qflgs. Can be dynamically resized
+typedef struct {
+  int *value;
+  size_t used;
+  size_t size;
+} qflgData;
+
+
+void free_parameters(struct RadarParm *prm, struct FitData *fit, FILE *fp, qflgData *q)
+{
+  if (prm != NULL) RadarParmFree(prm);
+  if (fit != NULL) FitFree(fit);
+  if (fp !=NULL) free(fp);
+  if (q->value !=NULL) {
+    free(q->value);
+    q->value=NULL;
+  }
+}
+
+
+
+int main (int argc,char *argv[]) {
+  
+  unsigned char vb=0;
+  unsigned char help=0;
+  unsigned char option=0;
+  unsigned char version=0;
+
+  OptionAdd(&opt,"-help",'x',&help);
+  OptionAdd(&opt,"-option",'x',&option);
+  OptionAdd(&opt,"-version",'x',&version);
+  OptionAdd(&opt,"vb",'x',&vb);
+
+  int arg;
+  arg=OptionProcess(1,argc,argv,&opt,rst_opterr);
+
+  if (arg==-1) {
+    exit(-1);
+  }
+
+  if (help==1) {
+    OptionPrintInfo(stdout,hlpstr);
+    exit(0);
+  }
+  if (option==1) {
+    OptionDump(stdout,&opt);
+    exit(0);
+  }
+
+  if (version==1) {
+    OptionVersion(stdout);
+    exit(0);
+  }
+  
+  FILE *fp=NULL;
+  if (arg==argc) fp=stdin;
+  else fp=fopen(argv[arg],"r");
+  if (fp==NULL) {
+    fprintf(stderr,"File not found.\n");
+    exit(-1);
```

Comment:
Wanna try using errno for the library? 

https://www.tutorialspoint.com/cprogramming/c_error_handling.htm
https://www-numi.fnal.gov/offline_software/srt_public_context/WebDocs/Errors/unix_system_errors.html